src/lib/crypto/kimchi_bindings/wasm/src/srs.rs

```diff
@@ -134,11 +134,14 @@ macro_rules! impl_srs {
                     JsValue::from_str("caml_pasta_fp_urs_lagrange_commitment")
                 })?;
 
-                let evals = (0..domain_size)
-                    .map(|j| if i == j { <$F as ark_ff::One>::one() } else { <$F as ark_ff::Zero>::zero() })
-                    .collect();
-                let p = Evaluations::<$F>::from_vec_and_domain(evals, x_domain).interpolate();
-                Ok(srs.commit_non_hiding(&p, None).into())
+                {
+                    // We're single-threaded, so it's safe to grab this pointer as mutable.
+                    // Do not try this at home.
+                    let srs = unsafe { &mut *((&**srs as *const SRS<$G>) as *mut SRS<$G>) as &mut SRS<$G> };
```

Comment:
so IIUC we're choosing to cache lagrange basis on demand here

Reply:
We still have to do the unsafe conversion AFAICT, since there are multiple non-mutable references to the SRS. This is the only way I could get it to work.

Reply:
a better way would be to use a [RwLock](https://doc.rust-lang.org/std/sync/struct.RwLock.html) imo